File: donater-bot/handlers/admin/reviews_mod.py
import logging

from aiogram import F, Router
from aiogram.fsm.context import FSMContext
from aiogram.types import CallbackQuery, Message

from config import config
from database import queries
from filters.admin import AdminFilter
from states.order import AdminFlow
from utils.formatter import escape_html

log = logging.getLogger(__name__)
router = Router(name="admin_reviews_mod")


def _tags_for_game(name: str) -> str:
    n = (name or "").lower()
    if "free" in n or "fire" in n:
        return "#Выполнено #FreeFire"
    if "pubg" in n:
        return "#Выполнено #PUBGMobile"
    return "#Выполнено"


@router.callback_query(F.data.regexp(r"^revok:(\d+)$"), AdminFilter())
async def cb_revok(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    rid = int((query.data or "").split(":")[-1])
    rev = await queries.review_by_id(rid)
    if not rev or rev["status"] != "pending":
        return
    oid = rev["order_id"]
    order = await queries.order_by_id(oid) if oid else None
    game = order["game_name"] if order else ""
    text = (
        f"{escape_html(rev['text'])}\n\n{_tags_for_game(game)}"
    )
    ch = config.reviews_channel.strip()
    if not ch:
        await query.message.answer("REVIEWS_CHANNEL не задан.")
        return
    try:
        msg = await query.bot.send_message(ch, text, parse_mode="HTML")
        await queries.review_set_status(rid, "published", msg.message_id)
    except Exception as exc:
        log.warning("publish review: %s", exc)
        await query.message.answer("Ошибка публикации.")
        return
    await query.message.answer("✅ Опубликовано.")


@router.callback_query(F.data.regexp(r"^revno:(\d+)$"), AdminFilter())
async def cb_revno(query: CallbackQuery, t):
    await query.answer()
    rid = int((query.data or "").split(":")[-1])
    await queries.review_set_status(rid, "rejected")
    await query.message.answer("Отзыв отклонён.")


@router.callback_query(F.data.regexp(r"^reved:(\d+)$"), AdminFilter())
async def cb_reved(query: CallbackQuery, t, state: FSMContext):
    await query.answer()
    rid = int((query.data or "").split(":")[-1])
    await state.set_state(AdminFlow.review_edit)
    await state.update_data(review_edit_id=rid)
    await query.message.answer("Новый текст отзыва:")


@router.message(AdminFlow.review_edit, F.text, AdminFilter())
async def msg_rev_edit(message: Message, t, state: FSMContext):
    data = await state.get_data()
    rid = data.get("review_edit_id")
    await state.clear()
    if not rid:
        return
    await queries.review_update_text(rid, (message.text or "").strip()[:2000])
    await message.answer("Текст обновлён (статус всё ещё pending — опубликуйте снова).")

File: donater-bot/handlers/admin/reviews_mod.js
import { Composer } from 'grammy';
import { reviewById, reviewSetStatus, reviewUpdateText, orderById } from '../../database/db.js';
import { config } from '../../config.js';
import { escapeHtml } from '../../utils/formatter.js';

export const adminReviewsComposer = new Composer();

adminReviewsComposer.callbackQuery(/^revok:(\d+)$/, async (ctx) => {
  try {
    await ctx.answerCallbackQuery();
    const id = parseInt(ctx.match[1], 10);
    const r = reviewById(id);
    if (!r || r.status !== 'pending') return;
    const o = r.order_id ? orderById(r.order_id) : null;
    const un = o?.username || 'user';
    const game = o?.game_name || '—';
    const product = o?.product_label || '—';
    const body =
      `⭐ <b>Отзыв покупателя</b>\n` +
      `👤 @${escapeHtml(un)}\n` +
      `🎮 ${escapeHtml(game)} — ${escapeHtml(product)}\n\n` +
      `<i>${escapeHtml(r.text)}</i>\n\n` +
      `#Выполнено #Review`;
    const ch = config.reviewsChannel;
    if (!ch) {
      await ctx.reply('REVIEWS_CHANNEL not set');
      return;
    }
    const sent = await ctx.api.sendMessage(ch, body, { parse_mode: 'HTML' });
    reviewSetStatus(id, 'approved', sent.message_id);
    await ctx.reply(ctx.i18n('review_publish_prompt'));
  } catch (e) {
    console.error(e);
  }
});

adminReviewsComposer.callbackQuery(/^revno:(\d+)$/, async (ctx) => {
  try {
    await ctx.answerCallbackQuery();
    const id = parseInt(ctx.match[1], 10);
    reviewSetStatus(id, 'rejected', null);
    await ctx.reply(ctx.i18n('review_rejected'));
  } catch (e) {
    console.error(e);
  }
});

adminReviewsComposer.callbackQuery(/^reved:(\d+)$/, async (ctx) => {
  try {
    await ctx.answerCallbackQuery();
    const id = parseInt(ctx.match[1], 10);
    ctx.session.step = 'admin_review_edit';
    ctx.session.adminReviewEditId = id;
    await ctx.reply(ctx.i18n('review_edited_prompt'));
  } catch (e) {
    console.error(e);
  }
});

adminReviewsComposer.on('message:text', async (ctx, next) => {
  try {
    if (ctx.session.step !== 'admin_review_edit') return next();
    const id = ctx.session.adminReviewEditId;
    const text = (ctx.message.text || '').trim();
    ctx.session.step = null;
    ctx.session.adminReviewEditId = null;
    reviewUpdateText(id, text);
    await ctx.reply(`Review #${id} updated.\nUse approve button again if needed.`);
  } catch (e) {
    console.error(e);
  }
});
